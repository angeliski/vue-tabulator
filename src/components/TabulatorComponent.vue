<template>
  <div
    ref="table"
    class="tabulator"
  />
</template>


<script lang='ts'>
import {
  Component, Prop, Vue, Watch, Model,
} from 'vue-property-decorator';
import Tabulator from 'tabulator-tables';

import mergeWith from 'lodash.mergewith';
import { IntegrationOptions, UpdateStrategy } from '@/types';
import merge from '../utilities/merge';
import eventFactory from '../feature/event-factory';
import cellEvents from '../feature/events/cell-events';
import rowEvents from '../feature/events/row-events';


@Component({
  name: 'TabulatorComponent',
})
export default class TabulatorComponent extends Vue {
  @Model('change', { default: () => [] })
  public tableData?: Array<any>;

  private tabulatorInstance: Tabulator | null = null;

  get eventOptions(): Object {
    const events = eventFactory.bind(this);
    return { ...events(rowEvents), ...events(cellEvents) };
  }

  @Prop({ default: () => ({}) })
  private options?: Tabulator.Options;

  @Prop({ default: () => ({ updateStrategy: UpdateStrategy.DATA }) })
  private integration?: IntegrationOptions;

  private resolvedOptions: Tabulator.Options = {};

  public getInstance() {
    return this.tabulatorInstance;
  }

  private createTable() {
    this.tabulatorInstance = new Tabulator(
      (this.$refs.table as HTMLElement),
      this.resolvedOptions,
    );
  }

  @Watch('options', { deep: true })
  private updateOptions() {
    this.resolvedOptions = {
      ...mergeWith({}, this.eventOptions, this.options, merge),
      data: this.tableData,
    };

    this.createTable();
  }

  @Watch('tableData', { deep: true })
  private updateData() {
    if (this.tabulatorInstance) {
      if (this.integration && this.integration.updateStrategy === UpdateStrategy.REPLACE) {
        this.tabulatorInstance.replaceData(this.tableData);
      } else if (this.integration && this.tableData
      && this.integration.updateStrategy === UpdateStrategy.UPDATE) {
        this.tabulatorInstance.updateData(this.tableData);
      } else {
        this.tabulatorInstance.setData(this.tableData);
      }
    }
  }

  mounted() {
    this.updateOptions();
  }
}
</script>
