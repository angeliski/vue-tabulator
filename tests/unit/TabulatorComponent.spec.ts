import { shallowMount, mount } from '@vue/test-utils';
import Vue from 'vue';
import TabulatorComponent from '@/components/TabulatorComponent.vue';
import { getInstance } from './helpers';

describe('TabulatorComponent.vue', () => {
  describe('Init table', () => {
    test('renders correct based in the options', () => {
      const options = {
        columns: [
          {
            title: 'Name',
            field: 'name',
            sorter: 'string',
            width: 200,
            editor: true,
          },
        ],
      };

      const wrapper = shallowMount(TabulatorComponent, {
        propsData: {
          options,
        },
      });


      expect(wrapper.findAll('.tabulator-col-title').filter(item => item.text().match('Name')).length).toBe(1);
      expect(wrapper).toMatchSnapshot();
    });
  });
  describe('Watchers', () => {
    test('update the config should recreate the table', async () => {
      const options : Tabulator.Options = {
        columns: [
          {
            title: 'Name',
            field: 'name',
            sorter: 'string',
            width: 200,
            editor: true,
          },
        ],
      };

      const wrapper = mount(TabulatorComponent, {
        propsData: {
          options,
        },
      });


      const oldInstance = getInstance(wrapper);
      wrapper.setProps({
        options: {
          columns: [
            {
              title: 'Age',
              field: 'age',
            },
          ],
        },
      });
      await Vue.nextTick();
      const currentInstance = getInstance(wrapper);

      expect(oldInstance).not.toBe(currentInstance);
    });

    test('update the config should remove the old column the table', async () => {
      const columns : Tabulator.ColumnDefinition[] = [
        {
          title: 'Name',
          field: 'name',
          sorter: 'string',
          width: 200,
          editor: true,
        },
        {
          title: 'Age',
          field: 'age',
          sorter: 'string',
          width: 200,
          editor: true,
        },
      ];
      const options : Tabulator.Options = {
        columns,
      };

      const wrapper = mount(TabulatorComponent, {
        propsData: {
          options,
        },
      });


      const oldInstance = getInstance(wrapper);
      wrapper.setProps({
        options: {
          columns: [columns[0]],
        },
      });
      await Vue.nextTick();
      const currentInstance = getInstance(wrapper);


      expect(oldInstance.getColumns()).not.toBe(currentInstance.getColumns());

      expect(currentInstance.getColumns().length).toBe(1);
      expect(currentInstance.getColumns()[0].getDefinition().title).toBe('Name');
    });

    test('update the v-model not should recreate the table', () => {
      const options : Tabulator.Options = {
        columns: [
          {
            title: 'Name',
            field: 'name',
            sorter: 'string',
            width: 200,
            editor: true,
          },
        ],
      };

      const wrapper = mount(TabulatorComponent, {
        propsData: {
          options,
        },
      });


      const oldInstance = getInstance(wrapper);
      wrapper.setProps({
        tableData: [],
      });
      const currentInstance = getInstance(wrapper);

      expect(oldInstance).toBe(currentInstance);
    });
  });
});
