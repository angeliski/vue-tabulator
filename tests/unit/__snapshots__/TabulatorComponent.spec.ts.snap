// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabulatorComponent.vue Init table renders correct based in the options 1`] = `
<div class="tabulator" role="grid" tabulator-layout="fitData">
  <div class="tabulator-header" style="padding-right: 0px;">
    <div class="tabulator-headers" style="margin-left: 0px;">
      <div class="tabulator-col tabulator-sortable" role="columnheader" aria-sort="none" tabulator-field="name" style="min-width: 40px; width: 200px;" title="">
        <div class="tabulator-col-content">
          <div class="tabulator-col-title-holder">
            <div class="tabulator-col-title">Name</div>
            <div class="tabulator-col-sorter">
              <div class="tabulator-arrow"></div>
            </div>
          </div>
        </div>
        <div class="tabulator-col-resize-handle"></div>
        <div class="tabulator-col-resize-handle prev"></div>
      </div>
    </div>
    <div class="tabulator-frozen-rows-holder"></div>
  </div>
  <div class="tabulator-tableHolder" tabindex="0" style="height: 0px;">
    <div class="tabulator-table"></div>
  </div>
</div>
`;
